features/26-mfa/api.md

```diff
@@ -0,0 +1,436 @@
+    return;
+  }
+  // Handle any other error.
+}
+```
+
+```typescript
+// Present an UI to let the user to choose authenticator.
+// Suppose the user chose TOTP.
+
+// Present AN UI to let the user to optionally name the TOTP authenticator.
+
+const displayName = textInput.value;
+const { authenticatorID, secret } = await skygear.auth.mfa.createNewTOTP(displayName);
+
+// Present to the secret to the user and let the user
+// to add the secret to their TOTP application, such as Authy and Google Authenticator.
+
+// Present an UI to instruct the user to input the OTP.
+const otp = textInput.value;
+const { recoveryCodes } = await skygear.auth.mfa.activateTOTP(authenticatorID, otp);
+```
+
+```typescript
+// Present an UI to let the user to choose authenticator.
+// Suppose the user chose OOB.
+
+// Present an UI to instruct the user to input a phone number.
+const e164number = textInput.value;
+const { authenticatorID } = await skygear.auth.mfa.createNewOOB({ phone: e164number });
+
+// Present an UI to instruct the user to check SMS.
+
+// Present an UI to offer the user to trigger again in case the delivery failed.
+await triggerOOB(authenticatorID);
+
+// Present an UI to instruct the user to input the recevied code.
+const code = textInput.value;
+const { recoveryCodes } = await skygear.auth.mfa.activateOOB(authenticatorID, code);
+```
+
+```typescript
+// Check if recoveryCodes is present and present them to the user.
+// Ask the user to remember them and keep them safe.
+if (recoveryCodes) {
+  navigateToDisplayRecoveryCodeScreen(recoveryCodes);
+}
+```
+
+## Authentication
+
+```typescript
+try {
+  await skygear.auth.login("user@example.com", "password");
+} catch (e) {
+  // Re-raise other error.
+  if (!skygear.isMFARequiredError(e)) {
+    throw e;
+  }
+
+  let done = false;
+
+  const bearerToken = await getBearerToken();
+  if (bearerToken) {
+    try {
+      await skygear.auth.mfa.authenticateWithBearerToken(bearerToken);
+      done = true;
+    } catch (ee) {
+      if (!skygear.isMFARequiredError(ee)) {
+        throw ee;
+      }
+    }
+  }
+
+  if (done) {
+    return;
+  }
+
+  const authenticators = await skygear.auth.mfa.listAuthenticators();
+  // Present the authenticators to the user and let them choose
+  // which one they want to use.
+
+  // Suppose they chose OOB
+  await skygear.mfa.triggerOOB(authenticators[0].id);
+
+  // Present an UI to instruct the user to check SMS.
+  // Present an UI to offer the user to trigger resend.
+  // Present an UI to instruct the user to input the received code.
+  // Present an UI to offer the user to skip MFA for 30 days.
+
+  const code = textInput.value;
+  const { bearerToken } = await skygear.auth.mfa.authenticateWithOOB({ authenticatorID, code }, { requestBearerToken: true});
+
+  // Persist the bearer token
+  if (bearerToken) {
+    await storeBearerToken(bearerToken);
+  }
+}
+```
+
+The above example demonstrates the flow of authenitcation in details.
```

Comment:
Could we have two set of codes here?

1. Maybe REST APIs / some sort of pesudocode like yours now, to present how it works underlying or on REST API level (like the bearer token)
2. A JS sample code to show the expected client side SDK.

Reason is I think besides bearer token, there are multiple things we may want to provide at client sdk which affect developer experience heavily. (e.g. The secret tokens of OTP should also be in the form of QR code too)

Reply:
> Could we have two set of codes here?

OK. I will add it.

> Reason is I think besides bearer token, there are multiple things we may want to provide at client sdk which affect developer experience heavily. (e.g. The secret tokens of OTP should also be in the form of QR code too)

Display the secret in form of QR code should be done by another library. I think the document should definitely teach the developer how to do this for each SDK we support.